R/stringify_expression.R

```diff
@@ -1,66 +1,31 @@
 ## input is quote()'ed expression
 ## reprex() takes care of that or, to use directly:
 ## q <- quote({a + b})
-stringify_expression <- function(q) {
-  src <- attr(q, "srcref")
 
-  ## if input is, eg 1:5, there is no srcref
-  if (is.null(src)) {
-    return(deparse(q))
-  }
-
-  ## https://journal.r-project.org/archive/2010-2/RJournal_2010-2_Murdoch.pdf
-  ## Each element is a vector of 6 integers: (first line, first byte, last line,
-  ## last byte, first character, last character)
-  ## ^^ clearly not whole story because I see vectors of length 8
-  ## but seems to be true description of first 6 elements
-  coord <- do.call(rbind, lapply(src, `[`, c(1, 5, 3, 6)))
-  colnames(coord) <- c("first_line", "first_character",
-                       "last_line", "last_character")
-  n_statements <- nrow(coord)
-  ## each row holds coordinates of source for one statement
-
-  ## all statements seem to contain such a reference to same lines of source
-  lines <-  attr(src[[1]], "srcfile")$lines
-
-  ## make sure source is split into character vector anticipated by coord
-  if (!isTRUE(attr(src[[1]], "srcfile")$fixedNewlines)) {
-    lines <- strsplit(lines, split = "\n")[[1]]
-  }
+#' @importFrom utils head tail getSrcref
+stringify_expression <- function(x) {
 
-  lines <- lines[coord[1, "first_line"]:coord[n_statements, "last_line"]]
+  # Construct a new sourcref with the starts from the first expression and the
+  # ends from the last one.
+  first_src <- head(getSrcref(x), n = 1)[[1]]
+  last_src <- tail(getSrcref(x), n = 1)[[1]]
```

Comment:
Why do you mix `head()`/`tail()` and positional indexing here? I assume there's a good reason and you're guarding against something.

Reply:
I think we can remove it, if `getSrcref(x)` doesn't exist it will be `NULL`, and `NULL[[1]][[1]]` and `NULL[[length(NULL)]][[1]]` both return `NULL`, so it doesn't matter.

I used tend to use `head()` and tail because they work better with 0 length vectors

``` r
x <- list()
head(x, n = 1)
#> list()
tail(x, n = 1)
#> list()

x[[1]]
#> Error in x[[1]]: subscript out of bounds
x[[length(x)]]
#> Error in x[[length(x)]]: attempt to select less than one element in integerOneIndex
```

But either way I should have used `head()` and `tail()` throughout, or not at all...

Reply:
If there was an existing rlang dep, it would probably be better to use `pluck`.